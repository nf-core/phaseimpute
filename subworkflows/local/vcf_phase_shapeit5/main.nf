include { GLIMPSE2_CHUNK                         } from '../../../modules/nf-core/glimpse2/chunk'
include { SHAPEIT5_PHASECOMMON                   } from '../../../modules/nf-core/shapeit5/phasecommon'
include { SHAPEIT5_LIGATE                        } from '../../../modules/nf-core/shapeit5/ligate'
include { BCFTOOLS_INDEX as VCF_BCFTOOLS_INDEX_1 } from '../../../modules/nf-core/bcftools/index'
include { BCFTOOLS_INDEX as VCF_BCFTOOLS_INDEX_2 } from '../../../modules/nf-core/bcftools/index'

workflow VCF_PHASE_SHAPEIT5 {

    take:
    ch_vcf        // channel (mandatory): [ [id, chr], vcf, csi, pedigree ]
    ch_region     // channel (optional) : [ [chr, region], region ]
    ch_ref        // channel (optional) : [ [id, chr], ref, csi ]
    ch_scaffold   // channel (optional) : [ [id, chr], scaffold, csi ]
    ch_map        // channel (optional) : [ [chr], map]

    main:

    ch_versions = Channel.empty()

    // Make chunks with Glimpse2 (does not work with "sequential" mode)
    chunk_model = "recursive"

    // Chunk with Glimpse2
    ch_input_glimpse2 = ch_vcf
        .map{
            metaIC, vcf, csi, pedigree -> [metaIC.subMap("chr"), metaIC, vcf, csi]
        }
        .combine(ch_region.map{ metaCR, region -> [metaCR.subMap("chr"), region]}, by:0)
        .join(ch_map)
        .map{
            metaC, metaIC, vcf, csi, region, gmap -> [metaIC, vcf, csi, region, gmap]
        }
    GLIMPSE2_CHUNK ( ch_input_glimpse2, chunk_model )
    ch_versions = ch_versions.mix( GLIMPSE2_CHUNK.out.versions.first() )

    // Rearrange channels
    ch_chunks_glimpse2 = GLIMPSE2_CHUNK.out.chunk_chr
        .splitCsv(
            header: [
                'ID', 'Chr', 'RegionBuf', 'RegionCnk', 'WindowCm',
                'WindowMb', 'NbTotVariants', 'NbComVariants'
            ], sep: "\t", skip: 0
        )
        .map { metaIC, it -> [metaIC, it["RegionBuf"], it["RegionCnk"]]}

    ch_chunks_number = GLIMPSE2_CHUNK.out.chunk_chr
        .map { meta, chunk -> [meta.subMap("chr"), chunk.countLines().intValue()]}

    ch_phase_input = ch_vcf
        .combine(ch_chunks_glimpse2, by:0)
        .map{
            metaIC, vcf, csi, pedigree, regionbuf, regioncnk -> [metaIC.subMap("chr"), metaIC, vcf, csi, pedigree, regionbuf, regioncnk]
        }
        .combine(ch_map, by:0)
        .map { metaC, metaIC, vcf, index, pedigree, regionbuf, regioncnk, gmap ->
            [metaIC + [chunk: regioncnk], vcf, index, pedigree, regionbuf, gmap]
        }

    SHAPEIT5_PHASECOMMON (
        ch_phase_input, ch_ref,
        ch_scaffold
    )
    ch_versions = ch_versions.mix(SHAPEIT5_PHASECOMMON.out.versions.first())

    VCF_BCFTOOLS_INDEX_1(SHAPEIT5_PHASECOMMON.out.phased_variant)
    ch_versions = ch_versions.mix(VCF_BCFTOOLS_INDEX_1.out.versions.first())

    ch_ligate_input = SHAPEIT5_PHASECOMMON.out.phased_variant
        .join(VCF_BCFTOOLS_INDEX_1.out.csi, failOnMismatch:true, failOnDuplicate:true)
        .map{ meta, vcf, csi -> [meta.subMap("id", "chr"), [vcf, meta.chunk], csi]}
        .groupTuple()
        .map{ meta, vcf, csi ->
                [ meta,
                vcf
                    .sort { a, b ->
                        def aStart = a.last().split("-")[-1].toInteger()
                        def bStart = b.last().split("-")[-1].toInteger()
                        aStart <=> bStart
                    }
                    .collect{it.first()},
                csi]}

    SHAPEIT5_LIGATE(ch_ligate_input)
    ch_versions = ch_versions.mix(SHAPEIT5_LIGATE.out.versions.first())

    VCF_BCFTOOLS_INDEX_2(SHAPEIT5_LIGATE.out.merged_variants)
    ch_versions = ch_versions.mix(VCF_BCFTOOLS_INDEX_2.out.versions.first())

    ch_vcf_tbi_join = SHAPEIT5_LIGATE.out.merged_variants
        .join(VCF_BCFTOOLS_INDEX_2.out.csi)

    emit:
    vcf_tbi             = ch_vcf_tbi_join         // channel: [ [id, chr], vcf, csi ]
    versions            = ch_versions             // channel: [ versions.yml ]
}

