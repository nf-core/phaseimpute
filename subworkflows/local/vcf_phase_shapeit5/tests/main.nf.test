nextflow_workflow {

    name "Test Subworkflow VCF_PHASE_SHAPEIT5"
    script "../main.nf"

    config "./nextflow.config"

    workflow "VCF_PHASE_SHAPEIT5"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/vcf_phase_shapeit5"
    tag "vcf_phase_shapeit5"

    tag "glimpse2"
    tag "glimpse2/chunk"
    tag "shapeit5"
    tag "shapeit5/phasecommon"
    tag "shapeit5/ligate"
    tag "bcftools"
    tag "bcftools/index"

    test("Phase vcf with regions, no map, no ref, no scaffold") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [id: "1000GP", chr: "chr22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz.csi", checkIfExist:true),
                        []
                    ],
                    [
                        [id: "1000GP", chr: "chr21"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21.s.norel.vcf.gz.csi", checkIfExist:true),
                        []
                    ]
                )
                input[1] = Channel.of(
                    [[chr: "chr22", region:"chr22:16570000-16610000"], "chr22:16570000-16610000"],
                    [[chr: "chr21", region:"chr21:16570000-16610000"], "chr21:16570000-16610000"]
                )
                input[2] = Channel.of([[],[],[]]).collect()
                input[3] = Channel.of([[],[],[]]).collect()
                input[4] = Channel.of(
                    [[chr: "chr22"],[]],
                    [[chr: "chr21"], []]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.vcf_tbi.collect{
                    path(it[1]).vcf.summary
                    }).match("Phasing content")
                },
                { workflow.out.vcf_tbi.collect{
                    assert path(it[1]).vcf.phased
                }}
            )
        }
    }
    /* TODO: Fix this test with https://github.com/odelaneau/shapeit5/issues/96
    test("Phase vcf with regions, with map, no ref, no scaffold") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [id: "1000GP", chr: "chr22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz.csi", checkIfExist:true),
                        []
                    ],
                    [
                        [id: "1000GP", chr: "chr21"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21.s.norel.vcf.gz.csi", checkIfExist:true),
                        []
                    ]
                )
                input[1] = Channel.of(
                    [[chr: "chr22", region:"chr22:16570000-16610000"], "chr22:16570000-16610000"],
                    [[chr: "chr21", region:"chr21:16570000-16610000"], "chr21:16570000-16610000"]
                )
                input[2] = Channel.of([[],[],[]]).collect()
                input[3] = Channel.of([[],[],[]]).collect()
                input[4] = Channel.of(
                    [ [chr: "chr22"], file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38_22.map", checkIfExist:true)],
                    [ [chr: "chr21"], file(params.pipelines_testdata_base_path + "hum_data/reference_genome/GRCh38_21.map", checkIfExist:true)]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.vcf_tbi.collect{
                    path(it[1]).vcf.summary
                    }).match("Phasing content with map")
                }
            )
        }
    }*/
}
