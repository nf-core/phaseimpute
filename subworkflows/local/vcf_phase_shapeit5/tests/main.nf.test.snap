{
    "Phasing content": {
        "content": [
            [
                "VcfFile [chromosomes=[chr21], sampleCount=3196, variantCount=836, phased=true, phasedAutodetect=true]",
                "VcfFile [chromosomes=[chr22], sampleCount=3196, variantCount=903, phased=true, phasedAutodetect=true]"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T15:56:20.067633938"
    },
    "Phase vcf with regions, no map, no ref, no scaffold": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "1000GP.vcf.gz:md5,a8af1c991fb1ca9e0919e0d5ddae8968",
                        "1000GP.vcf.gz.csi:md5,b4c653bd398bb838b496e022705bb562"
                    ],
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr22"
                        },
                        "1000GP.vcf.gz:md5,106157dc553c3c19eebdce0567e69d4f",
                        "1000GP.vcf.gz.csi:md5,58026cf302bd82382a078c5690429b5c"
                    ]
                ],
                "1": [
                    "versions.yml:md5,529c03b8d921c72026e91d71c0321811",
                    "versions.yml:md5,714bb0db6e2d39cf0042359a64915bc6",
                    "versions.yml:md5,8ffcda8a9d22f60c90d0e4276da3e714",
                    "versions.yml:md5,cab7592ebcb3d391afcd3191a175723b",
                    "versions.yml:md5,ed131d2608f28f8ada06ccc42717575e"
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr21"
                        },
                        "1000GP.vcf.gz:md5,a8af1c991fb1ca9e0919e0d5ddae8968",
                        "1000GP.vcf.gz.csi:md5,b4c653bd398bb838b496e022705bb562"
                    ],
                    [
                        {
                            "id": "1000GP",
                            "chr": "chr22"
                        },
                        "1000GP.vcf.gz:md5,106157dc553c3c19eebdce0567e69d4f",
                        "1000GP.vcf.gz.csi:md5,58026cf302bd82382a078c5690429b5c"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,529c03b8d921c72026e91d71c0321811",
                    "versions.yml:md5,714bb0db6e2d39cf0042359a64915bc6",
                    "versions.yml:md5,8ffcda8a9d22f60c90d0e4276da3e714",
                    "versions.yml:md5,cab7592ebcb3d391afcd3191a175723b",
                    "versions.yml:md5,ed131d2608f28f8ada06ccc42717575e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-22T15:46:34.727932091"
    }
}